// @temporalio/client/lib/types.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.decodePendingActivityState = exports.encodePendingActivityState = exports.PendingActivityState = exports.decodeActivityExecutionStatus = exports.encodeActivityExecutionStatus = exports.ActivityExecutionStatus = exports.decodeActivityIdConflictPolicy = exports.encodeActivityIdConflictPolicy = exports.ActivityIdConflictPolicy = exports.decodeActivityIdReusePolicy = exports.encodeActivityIdReusePolicy = exports.ActivityIdReusePolicy = exports.decodeQueryRejectCondition = exports.encodeQueryRejectCondition = exports.QueryRejectCondition = exports.InternalConnectionLikeSymbol = exports.HealthService = exports.TestService = exports.OperatorService = exports.WorkflowService = void 0;
const internal_workflow_1 = require("@temporalio/common/lib/internal-workflow");
const proto = __importStar(require("@temporalio/proto"));
exports.WorkflowService = proto.temporal.api.workflowservice.v1.WorkflowService;
exports.OperatorService = proto.temporal.api.operatorservice.v1.OperatorService;
exports.TestService = proto.temporal.api.testservice.v1.TestService;
exports.HealthService = proto.grpc.health.v1.Health;
exports.InternalConnectionLikeSymbol = Symbol('__temporal_internal_connection_like');
exports.QueryRejectCondition = {
    NONE: 'NONE',
    NOT_OPEN: 'NOT_OPEN',
    NOT_COMPLETED_CLEANLY: 'NOT_COMPLETED_CLEANLY',
    /** @deprecated Use {@link NONE} instead. */
    QUERY_REJECT_CONDITION_NONE: 'NONE',
    /** @deprecated Use {@link NOT_OPEN} instead. */
    QUERY_REJECT_CONDITION_NOT_OPEN: 'NOT_OPEN',
    /** @deprecated Use {@link NOT_COMPLETED_CLEANLY} instead. */
    QUERY_REJECT_CONDITION_NOT_COMPLETED_CLEANLY: 'NOT_COMPLETED_CLEANLY',
    /** @deprecated Use `undefined` instead. */
    QUERY_REJECT_CONDITION_UNSPECIFIED: undefined,
};
_a = (0, internal_workflow_1.makeProtoEnumConverters)({
    [exports.QueryRejectCondition.NONE]: 1,
    [exports.QueryRejectCondition.NOT_OPEN]: 2,
    [exports.QueryRejectCondition.NOT_COMPLETED_CLEANLY]: 3,
    UNSPECIFIED: 0,
}, 'QUERY_REJECT_CONDITION_'), exports.encodeQueryRejectCondition = _a[0], exports.decodeQueryRejectCondition = _a[1];
/**
 * @experimental Standalone Activities are experimental. APIs may be subject to change.
 */
exports.ActivityIdReusePolicy = {
    ALLOW_DUPLICATE: 'ALLOW_DUPLICATE',
    ALLOW_DUPLICATE_FAILED_ONLY: 'ALLOW_DUPLICATE_FAILED_ONLY',
    REJECT_DUPLICATE: 'REJECT_DUPLICATE',
};
/**
 * @experimental Standalone Activities are experimental. APIs may be subject to change.
 */
_b = (0, internal_workflow_1.makeProtoEnumConverters)({
    [exports.ActivityIdReusePolicy.ALLOW_DUPLICATE]: 1,
    [exports.ActivityIdReusePolicy.ALLOW_DUPLICATE_FAILED_ONLY]: 2,
    [exports.ActivityIdReusePolicy.REJECT_DUPLICATE]: 3,
    UNSPECIFIED: 0,
}, 'ACTIVITY_ID_REUSE_POLICY_'), exports.encodeActivityIdReusePolicy = _b[0], exports.decodeActivityIdReusePolicy = _b[1];
/**
 * @experimental Standalone Activities are experimental. APIs may be subject to change.
 */
exports.ActivityIdConflictPolicy = {
    FAIL: 'FAIL',
    USE_EXISTING: 'USE_EXISTING',
};
/**
 * @experimental Standalone Activities are experimental. APIs may be subject to change.
 */
_c = (0, internal_workflow_1.makeProtoEnumConverters)({
    [exports.ActivityIdConflictPolicy.FAIL]: 1,
    [exports.ActivityIdConflictPolicy.USE_EXISTING]: 2,
    UNSPECIFIED: 0,
}, 'ACTIVITY_ID_CONFLICT_POLICY_'), exports.encodeActivityIdConflictPolicy = _c[0], exports.decodeActivityIdConflictPolicy = _c[1];
/**
 * @experimental Standalone Activities are experimental. APIs may be subject to change.
 */
exports.ActivityExecutionStatus = {
    RUNNING: 'RUNNING',
    COMPLETED: 'COMPLETED',
    FAILED: 'FAILED',
    CANCELED: 'CANCELED',
    TERMINATED: 'TERMINATED',
    TIMED_OUT: 'TIMED_OUT',
};
/**
 * @experimental Standalone Activities are experimental. APIs may be subject to change.
 */
_d = (0, internal_workflow_1.makeProtoEnumConverters)({
    [exports.ActivityExecutionStatus.RUNNING]: 1,
    [exports.ActivityExecutionStatus.COMPLETED]: 2,
    [exports.ActivityExecutionStatus.FAILED]: 3,
    [exports.ActivityExecutionStatus.CANCELED]: 4,
    [exports.ActivityExecutionStatus.TERMINATED]: 5,
    [exports.ActivityExecutionStatus.TIMED_OUT]: 6,
    UNSPECIFIED: 0,
}, 'ACTIVITY_EXECUTION_STATUS_'), exports.encodeActivityExecutionStatus = _d[0], exports.decodeActivityExecutionStatus = _d[1];
/**
 * @experimental Standalone Activities are experimental. APIs may be subject to change.
 */
exports.PendingActivityState = {
    SCHEDULED: 'SCHEDULED',
    STARTED: 'STARTED',
    CANCEL_REQUESTED: 'CANCEL_REQUESTED',
    PAUSED: 'PAUSED',
    PAUSE_REQUESTED: 'PAUSE_REQUESTED',
};
/**
 * @experimental Standalone Activities are experimental. APIs may be subject to change.
 */
_e = (0, internal_workflow_1.makeProtoEnumConverters)({
    [exports.PendingActivityState.SCHEDULED]: 1,
    [exports.PendingActivityState.STARTED]: 2,
    [exports.PendingActivityState.CANCEL_REQUESTED]: 3,
    [exports.PendingActivityState.PAUSED]: 4,
    [exports.PendingActivityState.PAUSE_REQUESTED]: 5,
    UNSPECIFIED: 0,
}, 'PENDING_ACTIVITY_STATE_'), exports.encodePendingActivityState = _e[0], exports.decodePendingActivityState = _e[1];
